haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -27,42 +397,313 @@ import qualified Concordium.Crypto.SHA256 as Hash
+unOParam :: (IsSupported pt cpv ~ 'True) => OParam pt cpv a -> a
+unOParam (SomeParam a) = a
+
+-- |Lens for accessing the contents of an 'OParam' when the parameter is supported.
+supportedOParam :: (IsSupported pt cpv ~ 'True) => Lens' (OParam pt cpv a) a
+supportedOParam f (SomeParam a) = SomeParam <$> f a
+
+instance Functor (OParam pt cpv) where
+    fmap _ NoParam = NoParam
+    fmap f (SomeParam v) = SomeParam (f v)
+
+instance Foldable (OParam pt cpv) where
+    foldr _ b NoParam = b
+    foldr f b (SomeParam a) = f a b
+
+    foldl _ b NoParam = b
+    foldl f b (SomeParam a) = f b a
+
+    foldMap _ NoParam = mempty
+    foldMap f (SomeParam a) = f a
+
+instance Traversable (OParam pt cpv) where
+    traverse _ NoParam = pure NoParam
+    traverse f (SomeParam a) = SomeParam <$> f a
+
+instance (Eq a) => Eq (OParam pt cpv a) where
+    NoParam == NoParam = True
+    SomeParam a == SomeParam b = a == b
+
+instance (Ord a) => Ord (OParam pt cpv a) where
+    compare NoParam NoParam = EQ
+    compare (SomeParam a) (SomeParam b) = compare a b
+
+instance (Show a) => Show (OParam pt cpv a) where
+    show NoParam = "<parameter type unsupported>"
+    show (SomeParam a) = show a
+
+instance (Serialize a, SingI pt, IsChainParametersVersion cpv) => Serialize (OParam pt cpv a) where
+    put NoParam = return ()
+    put (SomeParam a) = put a
+
+    get = whenSupported get
+
+-- |Perform an action conditionally on whether the parameter is supported in the relevant chain
+-- parameters version (per 'sIsSupported'). The action is not performed if the parameter is not
+-- supported.
+whenSupported :: forall pt cpv f a. (Applicative f, SingI pt, IsChainParametersVersion cpv) => f a -> f (OParam pt cpv a)
+whenSupported m = case sIsSupported (sing @pt) (chainParametersVersion @cpv) of
+    SFalse -> pure NoParam
+    STrue -> SomeParam <$> m
+
+-- |Wrap a value in an 'OParam' depending on whether the parameter is supported.
+pureWhenSupported :: forall pt cpv a. (SingI pt, IsChainParametersVersion cpv) => a -> OParam pt cpv a
+pureWhenSupported v = case sIsSupported (sing @pt) (chainParametersVersion @cpv) of
+    SFalse -> NoParam
+    STrue -> SomeParam v
+
+-- |Analogue of 'maybe' for 'OParam'.
+maybeWhenSupported :: b -> (a -> b) -> OParam pt cpv a -> b
+maybeWhenSupported b _ NoParam = b
+maybeWhenSupported _ f (SomeParam a) = f a
+
+-- * Mint distribution
+
+-- |Constraint on a type level 'MintDistributionVersion' that can be used to get a corresponding
+-- 'SMintDistributionVersion'. An alias for 'SingI'.
+type IsMintDistributionVersion (mdv :: MintDistributionVersion) = SingI mdv
+
+-- |Witness a @SingI (SupportsMintPerSlot mdv)@ constraint using a 'SMintDistributionVersion mdv'.
+-- Concretely this provides the passed in action @a@ with the context 'SupportsMintPerSlot mdv'.
+withSupportsMintPerSlot :: forall mdv a. SMintDistributionVersion mdv -> (SingI (SupportsMintPerSlot mdv) => a) -> a
```

Comment:
Why is this one without an alias (i.e., why is it using SingI) explicitly.

Reply:
I have added an alias for a `SingI` i.e. ` type MintPerSlotSupported (mdv :: MintDistributionVersion) = SingI (SupportsMintPerSlot mdv) ` in order to make things consistent.